rosette/rbl/rosette/consume.rbl

```diff
@@ -0,0 +1,111 @@
+;;; File containing the reflective method consume and its helper methods
+
+(defOprn extract_data_candidates)
+(defProc (extract_data_candidates subspaces product_ptrns)
+    (map subspaces (proc [i [data_list continuation_struct_list]]
+        (map data_list (proc [data_index data]
+            (if (match_pattern (tuple-safe-nth product_ptrns i) data)
+                [data_index data]
+            ))))))
+
+;;; This function would have roughly below signature in Scala
+;;;
+;;; def select_best_products(candidates: List[List[Candidate]]): Option[List[Candidate]]
+;;;
+;;; For each channel, we choose the product to match with.
+;;; The input is a list of list of potential candidates and from each inner list, a candidate is chosen.
+;;; The function currently just takes the first candidate for each channel but in the future
+;;; you will be able to pass in a ranking algorithm (see https://rchain.atlassian.net/browse/ROS-329) that will select candidates.
+;;; If any channels are missing candidates, it returns an empty list as that implies a incomplete match.
+(defOprn select_best_products)
+(defProc (select_best_products candidates)
+    (let
+        [[prefiltered_reductions (map candidates (proc [i channel_candidates] (non_niv_head channel_candidates)))]]
+        (if (any_null? prefiltered_reductions) [] prefiltered_reductions)))
+
+;;; For each channel returned from the select_best_products function,
+;;; delete the matched products from the subspace and return a copy of the products.
+(defOprn consume_products)
+(defProc (consume_products reductions chart channels)
+    (map reductions (proc [i chosen_candidate]
+        (let*
+            [
+                [[data_index data] chosen_candidate]
+                [channel (tuple-safe-nth channels i)]
+                [[data_list continuation_struct_list] (tbl-get chart channel)]
+                [updated_data_list (delete-i data_list data_index)]
+            ]
+            (seq
+                (tbl-add chart channel [updated_data_list continuation_struct_list])
+                data
+            )))))
+
```

Comment:
This file is really hard to read:

  http://people.ace.ed.ac.uk/staff/medward2/class/moz/cm/doc/contrib/lispstyle.html

  https://stackoverflow.com/questions/541407/lisp-code-formatting

  http://dept-info.labri.fr/~strandh/Teaching/PFS/Common/Strandh-Tutorial/indentation.html